src/lib.rs

```diff
@@ -60,98 +41,251 @@ fn set_stack_limit(l: usize) {
+            };
+            let result = unsafe {
+                libc::mprotect((map as usize + page_size) as *mut libc::c_void,
+                               stack_size - page_size,
+                               libc::PROT_READ | libc::PROT_WRITE)
+            };
+            if result == -1 {
+                panic!("unable to set stack permissions")
+            }
+            let stack_low = map as usize;
+
+            // Prepare stack limits for the stack switch
+            set_stack_limit(stack_low);
+
+            // Make sure the stack is 16-byte aligned which should be enough for all
+            // platforms right now. Allocations on 64-bit are already 16-byte aligned
+            // and our switching routine doesn't push any other data, but the routine on
+            // 32-bit pushes an argument so we need a bit of an offset to get it 16-byte
+            // aligned when the call is made.
+            let offset = if cfg!(target_pointer_width = "32") {
+                12
+            } else {
+                0
+            };
+
+            unsafe {
+                __stacker_switch_stacks(&mut f as *mut &mut FnMut() as *mut u8,
+                                        doit as usize as *const _,
+                                        stack_low + stack_size - offset);
+            }
+
+            // Dropping `switch` frees the memory mapping and restores the old stack limit
+        }
     }
 }
 
+extern fn doit(f: &mut &mut FnMut()) {
+    f();
+}
+
 cfg_if! {
     if #[cfg(windows)] {
-        // See this for where all this logic is coming from.
-        //
-        // https://github.com/adobe/webkit/blob/0441266/Source/WTF/wtf
-        //                   /StackBounds.cpp
-        unsafe fn guess_os_stack_limit() -> usize {
-            #[cfg(target_pointer_width = "32")]
-            extern {
-                #[link_name = "__stacker_get_tib_32"]
-                fn get_tib_address() -> *const usize;
+        extern {
+            fn __stacker_get_current_fiber() -> winapi::PVOID;
+        }
+
+        #[no_mangle]
+        pub unsafe extern fn __stacker_switch_stacks_callback(f: &mut &mut FnMut()) {
+            f();
+        }
+
+        struct FiberInfo<'a> {
+            callback: &'a mut FnMut(),
+            result: Option<std::thread::Result<()>>,
+            parent_fiber: winapi::LPVOID,
+        }
+
+        unsafe extern "system" fn fiber_proc(info: winapi::LPVOID) {
+            let info = &mut *(info as *mut FiberInfo);
+            info.result = Some(std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                (info.callback)();
+            })));
+            kernel32::SwitchToFiber(info.parent_fiber);
+            return;
+        }
+
+        fn _grow(stack_size: usize, callback: &mut FnMut()) {
+            unsafe {
+                let was_fiber = kernel32::IsThreadAFiber() == winapi::TRUE;
+                
+                let mut info = FiberInfo {
```

Comment:
Can some comments be added about what's going on here? I have no idea what windows fibers are :(

Reply:
They are basically stackfull coroutines.